### PR TITLE
pmt: Serialize PMTs to byte strings in Python (and vice versa)

### PR DESCRIPTION
**Pull request by mbr0wn**

### CODE REVIEW

**Comment by mbr0wn** [comment]
This is a suggestion for a workaround for the PMT serialization, and maybe we can use this elsewhere. I tried forcing a bytes string in Python 3, but failed. However, I could force a tuple and then convert that.

File: gnuradio-runtime/swig/pmt_swig.i [hunk 1]
```diff
@@ -302,7 +319,21 @@ namespace pmt{
...
 } //namespace pmt
+
+%pythoncode %{
+  def serialize_str(pmt_obj):
+    import array
+    tup = _serialize_str_u8(pmt_obj)
+    return array.array('B', tup).tobytes()
+
+  def deserialize_str(pmt_str):
+    tup = tuple([x for x in pmt_str])
```
Comment (noc0lour) [inline gnuradio-runtime/swig/pmt_swig.i]: it should be possible to drop `[]` and do the list comprehension inside `tuple()`

**Comment by noc0lour** [comment]
Looks like the py3k test `qa_pmt` works now, but fails on py2

File: gnuradio-runtime/swig/pmt_swig.i (hunk 1 above)
Comment (mbr0wn) [reply gnuradio-runtime/swig/pmt_swig.i]: I don't get it. There is a list comprehension inside `tuple()`.

Comment (noc0lour) [reply gnuradio-runtime/swig/pmt_swig.i]: But you should be able to do this `tuple(x for x in pmt_str)` as well (:

Comment (mbr0wn) [reply gnuradio-runtime/swig/pmt_swig.i]: Oh you mean *don't* do a list comprehension :grin:

Yeah you're right; in Py3k it's almost muscle memory to use the [] to force unrolling of iterators, but that `tuple()` will unroll the iterator as well.

Comment (noc0lour) [reply gnuradio-runtime/swig/pmt_swig.i]: Ok, sorry, I meant generator expressions https://www.python.org/dev/peps/pep-0289/

File: gnuradio-runtime/python/pmt/qa_pmt.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 #
-# Copyright 2011 Free Software Foundation, Inc.
+# Copyright 2011,2018 Free Software Foundation, Inc.
```
Comment (marcusmueller) [inline gnuradio-runtime/python/pmt/qa_pmt.py]: Feels good man

File: gnuradio-runtime/swig/pmt_swig.i [hunk 3]
```diff
@@ -302,7 +319,21 @@ namespace pmt{
...
+  std::vector<uint8_t> _serialize_str_u8(pmt_t obj);
+  pmt_t _deserialize_str_u8(std::vector<uint8_t> py_str);
 
 } //namespace pmt
+
+%pythoncode %{
+  def serialize_str(pmt_obj):
+    import array
+    tup = _serialize_str_u8(pmt_obj)
+    return array.array('B', tup).tobytes()
```
Comment (marcusmueller) [inline gnuradio-runtime/swig/pmt_swig.i]: I'm afraid `array.array.tobytes` is Py3-only.

the base problem here being that `bytes` in Py2 gives a string, whereas it's a separate type in Py3k. 
Hm. I don't have a portable solution myself :(

**Comment by mbr0wn** [comment]
OK this one works for me both for Py2 and 3.

**Pull request merged**